src/kvstore/Listener.cpp

```diff
@@ -0,0 +1,202 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "kvstore/Listener.h"
+#include "kvstore/LogEncoder.h"
+#include "codec/RowReaderWrapper.h"
+
+DEFINE_int32(listener_commit_interval_secs, 1, "Listener commit interval");
+DEFINE_int32(listener_commit_batch_size, 1000, "Max batch size when listener commit");
+
+namespace nebula {
+namespace kvstore {
+
+Listener::Listener(GraphSpaceID spaceId,
+                   PartitionID partId,
+                   HostAddr localAddr,
+                   const std::string& walPath,
+                   std::shared_ptr<folly::IOThreadPoolExecutor> ioPool,
+                   std::shared_ptr<thread::GenericThreadPool> workers,
+                   std::shared_ptr<folly::Executor> handlers,
+                   std::shared_ptr<raftex::SnapshotManager> snapshotMan,
+                   std::shared_ptr<RaftClient> clientMan,
+                   meta::SchemaManager* schemaMan)
+    : RaftPart(FLAGS_cluster_id, spaceId, partId, localAddr, walPath,
+               ioPool, workers, handlers, snapshotMan, clientMan)
+    , schemaMan_(schemaMan) {
+}
+
+void Listener::start(std::vector<HostAddr>&& peers, bool) {
+    std::lock_guard<std::mutex> g(raftLock_);
+
+    lastLogId_ = wal_->lastLogId();
+    lastLogTerm_ = wal_->lastLogTerm();
+    term_ = proposedTerm_ = lastLogTerm_;
+
+    // Set the quorum number
+    quorum_ = (peers.size() + 1) / 2;
+
+    auto logIdAndTerm = lastCommittedLogId();
+    committedLogId_ = logIdAndTerm.first;
+
+    if (lastLogId_ < committedLogId_) {
+        LOG(INFO) << idStr_ << "Reset lastLogId " << lastLogId_
+                << " to be the committedLogId " << committedLogId_;
+        lastLogId_ = committedLogId_;
+        lastLogTerm_ = term_;
+        wal_->reset();
+    }
+
+    lastApplyLogId_ = lastApplyLogId();
+
+    LOG(INFO) << idStr_ << "Listener start"
+                        << ", there are " << peers.size() << " peer hosts"
+                        << ", lastLogId " << lastLogId_
+                        << ", lastLogTerm " << lastLogTerm_
+                        << ", committedLogId " << committedLogId_
+                        << ", lastApplyLogId " << lastApplyLogId_
+                        << ", term " << term_;
+
+    // As for listener, we don't need Host actually. However, listener need to be aware of
+    // membership change, it can be handled in preProcessLog.
+    for (auto& addr : peers) {
+        auto hostPtr = std::make_shared<raftex::Host>(addr, shared_from_this());
+        hosts_.emplace_back(hostPtr);
+    }
+
+    status_ = Status::RUNNING;
+    role_ = Role::LEARNER;
+
+    size_t delayMS = 100 + folly::Random::rand32(900);
+    bgWorkers_->addDelayTask(delayMS, &Listener::doApply, this);
```

Comment:
A question，does it affect the next `doApply` if the last one has not been completed?

Reply:
Yeah, we use `addDelayTask` one by one to avoid affect the next `doApply`, if previous failed, the `lastApplyLogId_` won't be updated, next will retry again.